dlt/destinations/impl/lancedb/lancedb_client.py

```diff
@@ -0,0 +1,567 @@
+import uuid
+from types import TracebackType
+from typing import (
+    ClassVar,
+    List,
+    Any,
+    cast,
+    Union,
+    Tuple,
+    Iterable,
+    Type,
+    Optional,
+)
+
+import lancedb  # type: ignore
+import pyarrow as pa
+from lancedb import DBConnection
+from lancedb.common import DATA  # type: ignore
+from lancedb.embeddings import EmbeddingFunctionRegistry, TextEmbeddingFunction  # type: ignore
+from lancedb.pydantic import LanceModel  # type: ignore
+from lancedb.query import LanceQueryBuilder  # type: ignore
+from lancedb.table import Table  # type: ignore[import-untyped]
+from numpy import ndarray
+from pyarrow import Array, ChunkedArray
+from pydantic import create_model
+
+from dlt.common import json, pendulum, logger
+from dlt.common.destination import DestinationCapabilitiesContext
+from dlt.common.destination.exceptions import DestinationUndefinedEntity
+from dlt.common.destination.reference import (
+    JobClientBase,
+    WithStateSync,
+    LoadJob,
+    StorageSchemaInfo,
+    StateInfo,
+    TLoadJobState,
+)
+from dlt.common.schema import Schema, TTableSchema, TSchemaTables
+from dlt.common.schema.typing import TColumnType
+from dlt.common.schema.utils import get_columns_names_with_prop
+from dlt.common.storages import FileStorage
+from dlt.common.typing import DictStrAny
+from dlt.destinations.impl.lancedb import capabilities
+from dlt.destinations.impl.lancedb.configuration import (
+    LanceDBClientConfiguration,
+)
+from dlt.destinations.impl.lancedb.errors import lancedb_error
+from dlt.destinations.impl.lancedb.lancedb_adapter import VECTORIZE_HINT
+from dlt.destinations.impl.lancedb.schema import (
+    TLanceModel,
+    create_template_schema,
+    make_fields,
+)
+from dlt.destinations.impl.lancedb.utils import (
+    list_unique_identifiers,
+    generate_uuid,
+    set_non_standard_providers_environment_variables,
+)
+from dlt.destinations.job_impl import EmptyLoadJob
+from dlt.destinations.type_mapping import TypeMapper
+
+
+class LanceDBTypeMapper(TypeMapper):
+    sct_to_unbound_dbt = {
+        "text": pa.string(),
+        "double": pa.float64(),
+        "bool": pa.bool_(),
+        "timestamp": pa.timestamp("us", "UTC"),
+        "bigint": pa.int64(),
+        "binary": pa.binary(),
+        "decimal": pa.decimal128(38, 18),
+        "date": pa.date32(),
+        "time": pa.time64("us"),
+        "complex": pa.string(),
+        "wei": pa.float64(),
+    }
+
+    sct_to_dbt = {}
+
+    dbt_to_sct = {
+        pa.string(): "text",
+        pa.float64(): "double",
+        pa.bool_(): "bool",
+        pa.timestamp("us", "UTC"): "timestamp",
+        pa.int64(): "bigint",
+        pa.binary(): "binary",
+        pa.decimal128(38, 18): "decimal",
+        pa.date32(): "date",
+        pa.time64("us"): "time",
+    }
+
+
+class NullSchema(LanceModel):
+    pass
+
+
+class VersionSchema(LanceModel):
```

Comment:
do we need these models or can we not just load to those tables the same way the loadpackages are loaded?

Reply:
Ideally yes, but the loadpackages loading utilities have a bunch of extra stuff like "special" fields and rather intricate workings to convert to pydantic models.

Can I request for this to be put as a separate ticket? Its not immediately obvious to me how to create a single interface here